bundles-maven-plugin/src/main/java/org/apache/metron/BundleProvidedDependenciesMojo.java

```diff
@@ -0,0 +1,326 @@
+public class BundleProvidedDependenciesMojo extends AbstractMojo {
+
+    /**
+     * The Maven project.
+     */
+    @Parameter(defaultValue = "${project}", readonly = true, required = true)
+    private MavenProject project;
+
+    /**
+     * The local artifact repository.
+     */
+    @Parameter(defaultValue = "${localRepository}", readonly = true)
+    private ArtifactRepository localRepository;
+
+    /**
+     * The {@link RepositorySystemSession} used for obtaining the local and remote artifact repositories.
+     */
+    @Parameter(defaultValue = "${repositorySystemSession}", readonly = true)
+    private RepositorySystemSession repoSession;
+
+    /**
+     * If specified, this parameter will cause the dependency tree to be written using the specified format. Currently supported format are: <code>tree</code>
+     * or <code>pom</code>.
+     */
+    @Parameter(property = "mode", defaultValue = "tree")
+    private String mode;
+
+    /**
+     * The type we are using for dependencies, should be nar, but may
+     * be changed in the configuration if the plugin is producing
+     * other archive extensions, this is a 'shared' configuration
+     * with the NarMojo
+     */
+    @Parameter(property = "type", required = false, defaultValue = "bundle")
+    protected String type;
+
+    /**
+     * The dependency tree builder to use for verbose output.
+     */
+    @Component
+    private DependencyTreeBuilder dependencyTreeBuilder;
+
+    /**
+     * *
+     * The {@link ArtifactHandlerManager} into which any extension {@link ArtifactHandler} instances should have been injected when the extensions were loaded.
+     */
+    @Component
+    private ArtifactHandlerManager artifactHandlerManager;
+
+    /**
+     * The {@link ProjectBuilder} used to generate the {@link MavenProject} for the nar artifact the dependency tree is being generated for.
+     */
+    @Component
+    private ProjectBuilder projectBuilder;
+
+    /*
+     * @see org.apache.maven.plugin.Mojo#execute()
+     */
+    @Override
+    public void execute() throws MojoExecutionException, MojoFailureException {
+        try {
+            // find the bundle dependency
+            Artifact bundleArtifact = null;
+            for (final Artifact artifact : project.getDependencyArtifacts()) {
+                if (type.equals(artifact.getType())) {
+                    // ensure the project doesn't have two bundle dependencies
+                    if (bundleArtifact != null) {
+                        throw new MojoExecutionException("Project can only have one NAR dependency.");
+                    }
+
+                    // record the bundle dependency
+                    bundleArtifact = artifact;
+                }
+            }
+
+            // ensure there is a bundle dependency
+            if (bundleArtifact == null) {
+                throw new MojoExecutionException("Project does not have any NAR dependencies.");
```

Comment:
done.

So, since everything in nifi is a bundle, they allow you to pull in at most one other bundle into your classloader.   The way they use it is to have a bundle of common code, re-used and pulled into other bundles when they are loaded.  Like, hadoop common get's pulled into the hadoop processors and the hadoop controllers which are separate bundles